File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == nullptr){
            return nullptr;
        }

        if(p->val < root->val && q->val < root->val){
            TreeNode* leftans = lowestCommonAncestor(root->left,p,q);
            return leftans;
        }
        if(p->val > root->val && q->val > root->val){
            TreeNode* rightans = lowestCommonAncestor(root->right,p,q);
            return rightans;
        }
        if(p->val < root->val && q->val > root->val){
            return root;
        }
        if(p->val > root->val && q->val < root->val){
            return root;
        }
        return root;
    }
};

File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null){
            return root;
        }
        else if(p.val < root.val && q.val < root.val){
            TreeNode left = lowestCommonAncestor(root.left,p,q);
            return left;
        }
        else if(p.val > root.val && q.val > root.val){
            TreeNode right = lowestCommonAncestor(root.right,p,q);
            return right;
        }
        return root;
    }
}